### ISSUE TITLE
About MemEffAttention

### ISSUE THREAD

**Issue by myc634** [facebookresearch/vggt#34]
Hi!

Thanks for the great work! While I am playing with the inference code, I find you guys have prepared the code for flash-attention, but didn't turn on the flag and using it, is their any concern about it? I also find that their is no qk-norm in the MemEffAttention module here:
`https://github.com/facebookresearch/vggt/blob/fa5911cb6be0b979b3507564bccdb95144178b0f/vggt/layers/attention.py#L80`
I feel the inference is really expensive when inference with dense mode. Is their any plan to use MemEffAttention in the future update?

Thanks for your answer!

**Comment by jytime** [comment]
Hi  @myc634 , 

Thanks for your interest and feedback!

The MemEffAttention module was inherited from DINO v2, but we disabled it intentionally because it requires installing xFormers, which can be challenging and painful to compile. Our goal is to make the setup as straightforward as possible, ideally requiring only PyTorch as a dependency, so we decided against enabling MemEffAttention by default.

Currently, there are no plans to enable MemEffAttention. However, our implementation already supports Flash Attention V2 seamlessly, provided you're using PyTorch version 2.3 or higher, as explained in this PyTorch announcement: https://pytorch.org/blog/pytorch2-2/.

Flash Attention V2 is automatically activated when using:

https://github.com/facebookresearch/vggt/blob/fa5911cb6be0b979b3507564bccdb95144178b0f/vggt/layers/attention.py#L60-L67

Note that self.fused_attn defaults to True in our codebase.

If you're looking to use an even more efficient attention mechanism (such as Flash Attention V3), you can compile FlashAttV3 or xFormers in your environment. Afterward, you only need to replace:

`F.scaled_dot_product_attention(...)`

with the equivalent implementation provided by the compiled library.

Hope this helps!

**Comment by myc634** [comment]
Thanks for your quick reply!

I also run into some other problem here, I try to use the following code to get the camera pose and point cloud:

```
aggregated_tokens_list, ps_idx = self.vggt_model.aggregator((images + 1) * 0.5) # denormalize into [0, 1]

pose_enc = self.vggt_model.camera_head(aggregated_tokens_list)[-1]

extrinsic, intrinsic = pose_encoding_to_extri_intri(pose_enc, images.shape[-2:])

last_row = torch.tensor([0, 0, 0, 1], device=extrinsic.device, dtype=extrinsic.dtype)
last_row = last_row.view(1, 1, 1, 4).expand(extrinsic.shape[0], extrinsic.shape[1], 1, 4)
extrinsic = torch.cat([extrinsic, last_row], dim=2)  # Shape becomes [B, V, 4, 4]
point_map, point_conf = self.vggt_model.point_head(aggregated_tokens_list, (images + 1) * 0.5, ps_idx) # denormalize into [0, 1]
```
My understanding is that the extrinsic is the world-to-camera translation in OpenCV coordinate system, so I try to visualize the pose and pointcloud in open3d, but find the result is really weird and looks like this:

![Image](https://github.com/user-attachments/assets/2ff51863-eb92-4826-9ecf-02f426b7a4af)

Here is the code I use in open3d:

```
import open3d as o3d
import numpy as np

num_points = 10000
points = np.load("point_map.npy") 
colors = np.load("point_colors.npy")


colors_normalized = (colors + 1) / 2.0
colors_normalized = np.clip(colors_normalized, 0, 1) 


pcd = o3d.geometry.PointCloud()
pcd.points = o3d.utility.Vector3dVector(points)
pcd.colors = o3d.utility.Vector3dVector(colors_normalized)


def create_camera_frames(extrinsics):
    frames = []
    coord_scale = 0.5 
    

    transform_correction = np.array([
        [1, 0, 0, 0],
        [0, -1, 0, 0],  
        [0, 0, -1, 0],  
        [0, 0, 0, 1]
    ])
    
    for ext in extrinsics:

        corrected_ext = ext @ transform_correction
        

        frame = o3d.geometry.TriangleMesh.create_coordinate_frame(
            size=coord_scale,
            origin=[0, 0, 0]
        )

        frame.transform(corrected_ext)
        frames.append(frame)
    
    return frames


camera_extrinsics = np.load("extrinsics.npy")  


vis_objects = [pcd] + create_camera_frames(camera_extrinsics)


vis = o3d.visualization.Visualizer()
vis.create_window(window_name='Joint Visualization', width=1920, height=1080)
for obj in vis_objects:
    vis.add_geometry(obj)


render_opt = vis.get_render_option()
render_opt.point_size = 2.0          
render_opt.background_color = [0.1, 0.1, 0.1] 
render_opt.light_on = True    

ctr = vis.get_view_control()
ctr.set_zoom(0.6)                    
ctr.set_front([-1, -1, -1])         

vis.run() 
vis.destroy_window()

```
When I trying to dump the extrinsics, I inverse them into camera-to-world transform matrix. But I still find this weird result, could you please give me some hint/context about this probelm so I can debug? Many thanks!

**Comment by myc634** [comment]
Also, I find another bug at:

`https://github.com/facebookresearch/vggt/blob/497e1bade42029e964de0a03b2ac42cf41d61e6a/vggt/heads/dpt_head.py#L209`

It should be:
`x = x[:, frames_start_idx:frames_end_idx].contiguous()`
otherwise the incontiguous gpu memory will cause error in the following `view` operation when bs>1

**Comment by jytime** [comment]
Thanks I am going to update the .contiguous() in the code.

For visualisation, can you try below?

```
import open3d as o3d
import numpy as np

flip_yz = np.array([
    [1,  0,  0, 0],
    [0, -1,  0, 0],
    [0,  0, -1, 0],
    [0,  0,  0, 1]
], dtype=np.float64)

def create_camera_vis(extrinsic, size=0.5):
    # Invert the extrinsic (since OpenCV extrinsic is world->camera, 
    # but we want camera->world for visualization).
    transform = np.linalg.inv(extrinsic)

    # Flip Y and Z so that +Y is up (Open3D style) rather than +Y down (OpenCV style).
    transform = flip_yz @ transform

    # Create a coordinate frame and transform it
    coord_frame = o3d.geometry.TriangleMesh.create_coordinate_frame(size=size)
    coord_frame.transform(transform)
    return [coord_frame]
```

**Comment by myc634** [comment]
Sorry for bothering again! I try to follow your paper to reproduce the camera normalization part, but I find that the gt camera normalized my my code seems missing some rotation part compare to the VGGT prediction, here is the visualization:

I transform all the points and camera pose into the first camera coord system and I vis them looks like this, the first camera aligns well:

![Image](https://github.com/user-attachments/assets/6549621d-4d5b-45e2-86ad-c4bf4b70c515)

But when it comes to the second camera, the pose seems missing some rotation between the vggt coord system and the camera coord system normalized by myself, it looks like this:

![Image](https://github.com/user-attachments/assets/40c1914c-246a-4a5b-8b30-65be63b14f77)

Also, the overall point cloud seems have the same problem, here is the code I use for normalization:

```
def normalize_cameras_batch_withpts3d(batches):
    intrinsics, extrinsics, pts3ds, valid_masks = batches["intrinsics"], batches["extrinsics"], batches["pts3d"], batches["valid_mask"] 
    bs = extrinsics.shape[0]
    normalized_cameras_c2ws, normalized_pts3ds, first_camera_w2cs, scene_scales = [], [], [], []

    for b in range(bs):
        camera_c2w, pts3d, valid_mask = extrinsics[b], pts3ds[b], valid_masks[b]
        # First, express all c2ws and pts3d in the first camera coordinate system
        first_camera_w2c = camera_c2w[0].inverse().unsqueeze(0).repeat(pts3d.shape[0], 1, 1)

        transformed_camera_c2w = torch.bmm(first_camera_w2c, camera_c2w)
        # Second, compute the Euclidean distance of all pts3d as scene scale
        pts3d_homo = torch.cat([pts3d, torch.ones_like(pts3d[:, :, :, :1])], dim=-1)
        transformed_pts3d = torch.bmm(first_camera_w2c, pts3d_homo.flatten(1, 2).transpose(1, 2)).transpose(1, 2)[..., :3]

        scene_scale = torch.norm(transformed_pts3d.flatten(0, 1)[valid_mask.flatten(0, 2).bool()], dim=-1).mean()

        # Third, normalize the camera c2w and pts3d
        transformed_camera_c2w[:, :3, 3] = transformed_camera_c2w[:, :3, 3] / scene_scale
        transformed_pts3d = transformed_pts3d / scene_scale

        normalized_cameras_c2ws.append(transformed_camera_c2w)
        normalized_pts3ds.append(transformed_pts3d)
        first_camera_w2cs.append(first_camera_w2c[0: 1])
        scene_scales.append(scene_scale)

    return torch.stack(normalized_cameras_c2ws), torch.stack(normalized_pts3ds), torch.stack(first_camera_w2cs), torch.stack(scene_scales)
```

All the camera extrinsics here are expressd as the camera-to-world transform matrix in the opencv coordinate system.

Am I missing anything or did something wrong? It will be great if you could point it out or give me some hint! That will be great!

**Comment by myc634** [comment]
Sorry for bothering again!

It would be great if you could tell the reason that manually set `self.use_checkpoint = False` at here:
`https://github.com/facebookresearch/vggt/blob/ae1b9c87507ea53c111c0bc79c20fb601d6ac283/vggt/layers/vision_transformer.py#L97`

I read the paper and find you guys are using grad check pointing, but I couldn't find related code, will you release them in the official training code?

Thanks for your time answering this question! Your answers really helps!

**Comment by jytime** [comment]
oh we default self.use_checkpoint as False just because we have not released the training code. You need to do nothing but call 

`torch.utils.checkpoint`

to attention layers, and it will work.
